### PR TITLE
bpo-44686 replace unittest.mock._importer with pkgutil.resolve_name

### PR DESCRIPTION
**Pull request by graingert**
https://bugs.python.org/issue44686

Automerge-Triggered-By: GH:cjw296

### CODE REVIEW

**Comment by tirkarthi** [comment]
I think this deserves a discussion and a bpo. @cjw296 this uses `pkgutil.resolve_name` which is a new feature in 3.9 and the backport needs to accommodate the change if merged.

File: Lib/unittest/mock.py [hunk 1]
```diff
@@ -1583,8 +1565,7 @@ def _get_target(target):
     except (TypeError, ValueError):
         raise TypeError("Need a valid target to patch. You supplied: %r" %
                         (target,))
-    getter = lambda: _importer(target)
-    return getter, attribute
+    return partial(pkgutil.resolve_name, target), attribute
```
Comment (FFY00) [inline Lib/unittest/mock.py:1596]: Is there any reason here to be using `functools.partial` over `lambda`?

**Comment by cjw296** [comment]
@tirkarthi: Looking at the diff, we could probably just skip backporting this one...

File: Lib/unittest/mock.py (hunk 1 above)
Comment (graingert) [reply Lib/unittest/mock.py:1596]: for consistency. `lambda:` is used twice in this file whereas `partial(` is used six times

**Comment by graingert** [comment]
> I think this deserves a discussion and a bpo. @cjw296 this uses `pkgutil.resolve_name` which is a new feature in 3.9 and the backport needs to accommodate the change if merged.

I've now created the bpo

**Review by cjw296 — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2021-07-20-19-35-49.bpo-44686.ucCGhu.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+replace unittest.mock._importer with pkgutil.resolve_name
```
Comment (cjw296) [inline Misc/NEWS.d/next/Library/2021-07-20-19-35-49.bpo-44686.ucCGhu.rst]:
```suggestion
Replace ``unittest.mock._importer`` with ``pkgutil.resolve_name``.

**Comment by J08nY** [comment]
Hi all, sorry to resurrect a very old PR. However, this change actually changed the behavior of the lookup used in the mockers. A clear example is given here: https://github.com/scikit-hep/pyhf/issues/2143#issuecomment-1647467653.

I wonder if the Python docs for unittest could be changed to include information about the changed behavior. However, I am not sure how to describe the actual changes.

**Pull request merged**